cli/src/main/kotlin/commands/DownloaderCommand.kt

```diff
@@ -177,48 +177,57 @@ class DownloaderCommand : CliktCommand(name = "download", help = "Fetch source c
 
         val errorMessages = mutableListOf<String>()
         packages.forEach { pkg ->
-            try {
-                val result = Downloader.download(pkg, File(outputDir, pkg.id.toPath()), allowMovingRevisions)
-
-                if (archive) {
-                    val zipFile = File(
-                        outputDir,
-                        "${pkg.id.type.encodeOrUnknown()}-${pkg.id.namespace.encodeOrUnknown()}-" +
-                                "${pkg.id.name.encodeOrUnknown()}-${pkg.id.version.encodeOrUnknown()}.zip"
-                    )
+            download(pkg, File(outputDir, pkg.id.toPath()), allowMovingRevisions, errorMessages)
+        }
 
-                    log.info {
-                        "Archiving directory '${result.downloadDirectory.absolutePath}' to " +
-                                "'${zipFile.absolutePath}'."
-                    }
+        if (errorMessages.isNotEmpty()) {
+            log.error { "Error Summary:\n\n${errorMessages.joinToString("\n\n")}" }
+            throw ProgramResult(2)
+        }
+    }
 
-                    try {
-                        result.downloadDirectory.packZip(
-                            zipFile,
-                            "${pkg.id.name.encodeOrUnknown()}/${pkg.id.version.encodeOrUnknown()}/"
-                        )
-                    } catch (e: IllegalArgumentException) {
-                        e.showStackTrace()
-
-                        log.error { "Could not archive '${pkg.id.toCoordinates()}': ${e.collectMessagesAsString()}" }
-                    } finally {
-                        val relativePath = outputDir.toPath().relativize(result.downloadDirectory.toPath()).first()
-                        File(outputDir, relativePath.toString()).safeDeleteRecursively()
-                    }
+    private fun download(
+        pkg: Package,
+        outputDir: File,
+        allowMovingRevisions: Boolean,
+        errorMessages: MutableList<String>
```

Comment:
Would it make sense to remove this `output parameter` in favor of returning the error messages?

Reply:
Depending on which commit you reviewed here, `download()` already returns a `DownloadResult`. Also, having `errorMessages` as an output parameter makes the collection of error messages in the caller slightly nicer.

Reply:
On higher abstraction level code (like this) I like designs without output parameters more, anyhow let's keep it as you prefer it. Consider adding the return value explicitly to the signature as this it is not very obvious as-is.

Reply:
> Consider adding the return value explicitly to the signature as this it is not very obvious as-is.

Done.

Reply:
Thanks for taking the time and listing the options above. Originally I favored the 2nd one you listed, but if that one does not fit well with the callers you mention let's leave it as-is. I don't see any other options either.